lark/lexer.py

```diff
@@ -409,25 +413,35 @@ def __init__(self, conf, states, always_accept=()):
         assert trad_conf.tokens is terminals
         self.root_lexer = TraditionalLexer(trad_conf)
 
-    def lex(self, stream, get_parser_state):
-        parser_state = get_parser_state()
-        l = _Lex(self.lexers[parser_state], parser_state)
+    def make_lexer_state(self, text):
+        return self.root_lexer.make_lexer_state(text)
+
+    def lex(self, lexer_state, parser_state):
         try:
-            for x in l.lex(stream, self.root_lexer.newline_types, self.root_lexer.ignore_types):
-                yield x
-                parser_state = get_parser_state()
-                l.lexer = self.lexers[parser_state]
-                l.state = parser_state # For debug only, no need to worry about multithreading
+            while True:
+                lexer = self.lexers[parser_state.position]
+                yield lexer.next_token(lexer_state)
+        except EOFError:
+            pass
         except UnexpectedCharacters as e:
             # In the contextual lexer, UnexpectedCharacters can mean that the terminal is defined,
             # but not in the current context.
             # This tests the input against the global context, to provide a nicer error.
-            root_match = self.root_lexer.match(stream, e.pos_in_stream)
+            root_match = self.root_lexer.match(lexer_state.text, e.pos_in_stream)
             if not root_match:
                 raise
 
             value, type_ = root_match
             t = Token(type_, value, e.pos_in_stream, e.line, e.column)
-            raise UnexpectedToken(t, e.allowed, state=e.state)
+            raise UnexpectedToken(t, e.allowed, state=parser_state.position)
+
+class LexerThread:
```

Comment:
Why not add a `.lexer` attribute to `LexerState`? This is also makes is somewhat harder to completely build a `LexerState` by hand for whatever reason.

Reply:
I had a similar thought, but came to realize that they should be separate entities.

That way `ContextualLexer` can pass the `LexerState` instance to `TraditionalLexer` without confusion.

If it has a `.lexer` attribute, then it would have to be sometimes used and sometimes ignored.